snuba/query/processors/prewhere.py

```diff
@@ -25,56 +34,189 @@ class PrewhereProcessor(QueryProcessor):
     Moves top level conditions into the pre-where clause
     according to the list of candidates provided by the query data source.
 
-    In order for a condition to become a pre-where condition it has
-    to be:
+    In order for a condition to become a pre-where condition it has to be:
     - a single top-level condition (not in an OR statement)
-    - any of its referenced columns must be in the list provided to the
-      constructor.
+    - any of its referenced columns must be in the list provided by the query
+      data source.
     """
 
     def __init__(self, max_prewhere_conditions: Optional[int] = None) -> None:
         self.__max_prewhere_conditions: Optional[int] = max_prewhere_conditions
 
-    def process_query(self, query: Query, request_settings: RequestSettings,) -> None:
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
         max_prewhere_conditions: int = (
             self.__max_prewhere_conditions or settings.MAX_PREWHERE_CONDITIONS
         )
         prewhere_keys = query.get_data_source().get_prewhere_candidates()
         if not prewhere_keys:
             return
-        prewhere_conditions: Sequence[Condition] = []
+
+        # While both implementations modify the query, they do not interfere with each
+        # other since one depend on the legacy representation and the other on the AST
+        # Thus we can execute the two independently.
+        LegacyPrewhereProcessor().process_query(
+            query, max_prewhere_conditions, prewhere_keys
+        )
+        ASTPrewhereProcessor().process_query(
+            query, max_prewhere_conditions, prewhere_keys
+        )
+
+
+TCondition = TypeVar("TCondition")
+TColumn = TypeVar("TColumn")
```

Comment:
Unfortunately I cannot achieve this with one type only since in the legacy implementation column and condition have no common parent class.